### PR TITLE
Remove hardcoded instruction handlers, automate instead.

### PR DESCRIPTION
**Pull request by dsharletg**
This PR removes the hardcoded if-else tree for getting the mask for instruction dependent Hexagon relocations. This has a few nice properties:

- A few of the old checks were slightly dangerous: for example, we used the test `(inst >> 24) == 65` to detect the following instructions:

    0100 0001 000s ssss PP0t tiii iiid dddd
    0100 0001 001s ssss PP0t tiii iiid dddd
    0100 0001 010s ssss PP0t tiii iiid dddd
    0100 0001 011s ssss PP0t tiii iiid dddd
    0100 0001 100s ssss PP0t tiii iiid dddd
    0100 0001 110s ssss PP0t tiii iiid dddd

This is dangerous because the check would theoretically match two more instructions that aren't currently defined, but might be in the future, and we don't know what mask those instructions would be given (for sanity's sake... I would hope the same one, but we don't know that).

- We shouldn't run into any more instructions we've never seen before, and have to go dig up the encoding and add a new check. This should cover all V60 instructions, including all of the instructions we previously supported.

The disadvantage to this approach is possibly (compile time, not run time) performance. This is probably why @abadams did it this way in the first place (this code was originally in the remote runtime, running on Hexagon, so this approach was probably not a good idea). However, I think it shouldn't matter now that this is running at compile time, even for large programs. If we need to, we can pre-process the encodings into masks to enable faster lookups. We could also theoretically consolidate encodings with the same mask (similar to how the old method covered many instructions with the same condition).

### CODE REVIEW

File: src/HexagonOffload.cpp [hunk 1]
```diff
@@ -224,6 +224,35 @@ std::string print_sections(const Object &obj) {
     return oss.str();
 }
 
+// NULL-terminated list of Hexagon instructions (defined at the end of
+// this file).
+extern const char *hexagon_instructions[];
+
+// Given an instruction and an encoding from hexagon_instructions,
+// check if the instruction inst is one of these encoded instructions,
```
Comment (khourig) [inline src/HexagonOffload.cpp]: inst?

**Comment by pranavb-ca** [comment]
Thanks for this PR. LGTM

**Comment by dsharletg** [comment]
I talked to Andrew about this, he said that some of the encodings in this list are wrong! So this may not be a safe change. @pranavb-ca @ronlieb assuming you want this change, maybe a way forward is to get a more authoritative instruction encoding list?

**Comment by abadams** [comment]
I'm trying to remember what the exact problem was. It wasn't that something
was just wrong, but I think it was sometimes the bits you want are the 'i'
bits, and sometimes they're the 'I' bits.

**Comment by pranavb-ca** [comment]
There are two ways we can go about this. One is to lay our hands on an authoritative list of encodings. The other is for us ( @ronlieb  and I) to check the feasibility of uploading some files in the LLVM hexagon backend that expose an API used by the linker to get the mask for an instruction.

**Comment by dsharletg** [comment]
I'm not opposed to either option, but there is a third choice: we can just keep using the system we're using today, adding instructions when we find them. This is annoying when it happens, but it's rare. I  also think the chances of any users running into this on their own is low, I think just about every instance of this occurring has been when we change the runtime. I think the set of instructions used by actual Halide pipeline code is pretty small and we've probably almost completely covered that set at this point.

**Comment by steven-johnson** [comment]
What's the status on this -- more work needed, or just abandoned?

**Comment by dsharletg** [comment]
Closing this. I think we should keep doing what we're doing, unless the LLVM implementation of this somehow gets exposed for us to use.

**Pull request closed without merging**